jerry-core/lit/lit-strings.c

```diff
@@ -31,89 +31,55 @@ lit_is_valid_utf8_string (const lit_utf8_byte_t *utf8_buf_p, /**< utf-8 string *
                           lit_utf8_size_t buf_size, /**< string size */
                           bool is_strict) /**< true if surrogate pairs are not allowed */
 {
-  lit_utf8_size_t idx = 0;
+  const unsigned char *end = buf_size + utf8_buf_p;
 
-  bool is_prev_code_point_high_surrogate = false;
-  while (idx < buf_size)
+  const unsigned char *idx = (const unsigned char *) utf8_buf_p;
+
+  while (idx < end)
   {
-    lit_utf8_byte_t c = utf8_buf_p[idx++];
-    if ((c & LIT_UTF8_1_BYTE_MASK) == LIT_UTF8_1_BYTE_MARKER)
+    const unsigned char c = *idx++;
+
+    if (c <= 0x7f)
     {
-      is_prev_code_point_high_surrogate = false;
       continue;
     }
 
-    lit_code_point_t code_point = 0;
-    lit_code_point_t min_code_point = 0;
-    lit_utf8_size_t extra_bytes_count;
-    if ((c & LIT_UTF8_2_BYTE_MASK) == LIT_UTF8_2_BYTE_MARKER)
-    {
-      extra_bytes_count = 1;
-      min_code_point = LIT_UTF8_2_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_5_BITS_MASK));
-    }
-    else if ((c & LIT_UTF8_3_BYTE_MASK) == LIT_UTF8_3_BYTE_MARKER)
-    {
-      extra_bytes_count = 2;
-      min_code_point = LIT_UTF8_3_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_4_BITS_MASK));
-    }
-    else if ((c & LIT_UTF8_4_BYTE_MASK) == LIT_UTF8_4_BYTE_MARKER)
-    {
-      extra_bytes_count = 3;
-      min_code_point = LIT_UTF8_4_BYTE_CODE_POINT_MIN;
-      code_point = ((uint32_t) (c & LIT_UTF8_LAST_3_BITS_MASK));
-    }
-    else
-    {
-      /* utf-8 string could not contain 5- and 6-byte sequences. */
-      return false;
-    }
+    const unsigned char d = *idx++;
 
-    if (idx + extra_bytes_count > buf_size)
+    if ((c & 0xe0) == 0xc0)
```

Comment:
`if (c <= 0xe1) return false` could simplify the check in line 51. Also `(d & 0xc0) != 0x80` check could be done here as well, since it is needed by all cases later. Furthermore a length check before `*idx++` is missing, which may lead to segfault.